server/app/socketService.js

```diff
@@ -0,0 +1,103 @@
+/* eslint-disable no-underscore-dangle */
+/* eslint-disable consistent-return */
+const SocketIo = require('socket.io');
+const SocketClient = require('socket.io-client');
+const httpServer = require('http').createServer();
+const User = require('./models/User');
+
+const ENDPOINT = process.env.SOCKET_CHANNEL_ENDPOINT || 'http://localhost';
+const PORT = process.env.SOCKET_CHANNEL_PORT || 3002;
+
+const io = SocketIo();
+const socketServer = {};
+
+/*
+ *  List of connected clients
+ *  {
+ *    [userId]: SocketObject
+ *  }
+ */
+socketServer.clients = {};
+
+socketServer.start = server => {
+  io.attach(server);
+  io.on('connection', socket => {
+    if (process.env.NODE_ENV !== 'production') {
+      console.log(`Socket: new connection from ${socket.id}`);
+    }
+
+    setTimeout(() => {
+      if (!socket.authenticated) return socket.disconnect(true);
+    }, 2000);
+
+    // Use user email to determin the identity of client
+    socket.on('authenticate', ({ email }) => {
+      if (!email) return socket.disconnect(true);
+
+      User.findOne({ email }, (err, user) => {
+        if (err) {
+          console.log(err);
+          return socket.disconnect(true);
+        }
+
+        if (!user) return socket.disconnect(true);
+
+        // eslint-disable-next-line no-param-reassign
+        socket.authenticated = true;
+
+        socket.on('disconnect', () => {
+          delete socketServer.clients[user._id];
+        });
+
+        // Add current connection to list of connected clients
+        socketServer.clients[user._id] = socket;
+
+        if (process.env.NODE_ENV !== 'production') {
+          console.log(Object.keys(socketServer.clients));
+        }
+      });
+    });
+  });
```

Comment:
Instead of passing the user's email, we can make use of the token in the cookie. I think it's more secure as the token cannot be accessed via JavaScript on the client.
```
    const cookie = require('cookie');
    const jwt = require('jsonwebtoken');
    ...
    // Use the token cookie to authenticate the client
    socket.on('authenticate', async () => {
      const { token } = cookie.parse(socket.handshake.headers.cookie);

      if (!token) {
        return socket.disconnect(true);
      }

      try {
        const decoded = jwt.verify(token, process.env.JWT_SECRET);
        const { id } = decoded;

        await User.findById(id);

        // eslint-disable-next-line no-param-reassign
        socket.authenticated = true;

        // Add current connection to list of connected clients
        socketServer.clients[id] = socket;

        if (process.env.NODE_ENV !== 'production') {
          console.log(Object.keys(socketServer.clients));
        }

        socket.on('disconnect', () => {
          delete socketServer.clients[id];
        });
      } catch (error) {
        console.log(error);
        return socket.disconnect(true);
      }
    });
```

Reply:
I think this can be done on the 'connect' event listener as well